Add harbor chart v1.7.5-rancher1 for global registry

gitlawr opened the pull request:
Related issue: https://github.com/rancher/rancher/issues/1738

Notes:
1. global registry is available in HA setup.
2. Expose registry `/v2` endpoint to the same domain of Rancher server.
3. Expose `/api/projects` endpoint to Harbor core. It may be needed when the global registry is used as another Harbor's replication target. it will try to create a project via this endpoint when it does not exist.
4. Use `/registry` endpoint to aggregate Harbor UI and its API/services endpoints.
5. Add a proxy layer to help Harbor work with path prefix.

gitlawr committed:
Add base harbor chart v1.7.5-rancher1

Co-authored-by: Guangbo <guangbo@rancher.com>
Co-authored-by: gitlawr <lawrleegle@gmail.com>

thxCode commented on charts/harbor/v0.0.1/templates/registry/registry-cm.yaml:
```diff
@@ -47,7 +47,7 @@ data:
         encrypt: {{ $storage.s3.encrypt }}
         {{- end }}
-        {{- if $storage.s3.secure }}
+        {{- if kindIs "bool" $storage.s3.secure }}
```
Why do we need `kindIs` here? When `$storage.s3.secure` is `false`, does it still return `true`?

gitlawr replied:
Because `true` is the default value for the `secure` option. 
Originally, if the value is false, the if block is skipped and the false value is not used.
Now, if the value is false, it is bool kind so the if condition is true.
`false` and nil is indistinguishable in `if` statement so `kindIs` function is used here.

thxCode replied:
This is not very intuitive to understand how to use `$storage.s3.secure`, could you add a new param to enter into this block?

gitlawr replied:
Updated

thxCode commented on charts/harbor/v0.0.1/templates/registry/registry-cm.yaml:
```diff
@@ -120,7 +120,7 @@ data:
         encrypt: {{ $storage.oss.encrypt }}
         {{- end }}
-        {{- if $storage.oss.secure }}
+        {{- if kindIs "bool" $storage.oss.secure }}
```
Why do we need `kindIs` here? When `$storage.oss.secure` is `false`, does it still return `true`?

gitlawr replied:
See above comment.

thxCode replied:
This is not very intuitive to understand how to use `$storage.oss.secure`, could you add a new param to enter into this block?

thxCode commented on charts/harbor/v0.0.1/templates/proxy/deployment.yaml:
```diff
@@ -0,0 +1,62 @@
+          - key: filters.conf
+            path: filters.conf
+    {{- with .Values.proxy.nodeSelector }}
+      nodeSelector:
```
Make sure all workloads don't be scheduled into Windows host. You could choose to use `node selector` or `node affinity`.

thxCode commented on charts/harbor/v0.0.1/templates/proxy/deployment.yaml:
```diff
@@ -0,0 +1,62 @@
+{{ toYaml . | indent 8 }}
+    {{- end }}
+    {{- with .Values.proxy.affinity }}
+      affinity:
```
Make sure all workloads don't be scheduled into Windows host. You could choose to use `node selector` or `node affinity`.

gitlawr replied:
Updated

thxCode reviewed (commented):
I don't know `registry` components, I'm just concerned with some usage issues. Everything else looks good, I gave some suggestions and doubts.

thxCode commented:
LGTM

gitlawr committed:
Add unit tests for harbor chart

gitlawr commented:
@thxCode Added unit tests for the chart, PTAL.

thxCode commented:
LGTM

gitlawr committed:
Add global registry support

Add globalRegistryMode to enable deploying the Harbor as a global
registry in Rancher. It use the same domain as Rancher server in
ingress rules, and use the same tls-rancher-ingress secret for TLS.

gitlawr commented:
Updated the chart version to `1.7.5-rancher1` per discussion

cjellick commented:
LGTM
